iterative/utils/analytics.go

```diff
@@ -0,0 +1,351 @@
+		return true
+	}
+
+	return len(guessCI()) > 0
+}
+
+func guessCI() string {
+	if _, ok := os.LookupEnv("GITHUB_SERVER_URL"); ok {
+		return "github"
+	}
+
+	if _, ok := os.LookupEnv("CI_SERVER_URL"); ok {
+		return "gitlab"
+	}
+
+	if _, ok := os.LookupEnv("BITBUCKET_WORKSPACE"); ok {
+		return "bitbucket"
+	}
+
+	if _, ok := os.LookupEnv("TF_BUILD"); ok {
+		return "azure"
+	}
+
+	return ""
+}
+
+func TerraformVersion() string {
+	var out bytes.Buffer
+	cmd := exec.Command("terraform", "--version")
+	cmd.Stdout = &out
+
+	err := cmd.Run()
+
+	if err == nil {
+		regex := regexp.MustCompile(`v\d+\.\d+\.\d+`)
+		version := regex.FindString(out.String())
+		return version
+	} else {
+		logrus.Error("Analytics: Failed extracting terraform version")
+	}
+
+	return ""
+}
+
+func GroupId() string {
+	if !IsCI() {
+		return ""
+	}
+
+	rawId := "CI"
+	ci := guessCI()
+
+	if ci == "github" {
+		rawId = fmt.Sprintf("%s/%s",
+			os.Getenv("GITHUB_SERVER_URL"),
+			os.Getenv("GITHUB_REPOSITORY_OWNER"))
+	} else if ci == "gitlab" {
+		rawId = fmt.Sprintf("%s/%s",
+			os.Getenv("CI_SERVER_URL"),
+			os.Getenv("CI_PROJECT_ROOT_NAMESPACE"))
+	} else if ci == "bitbucket" {
+		rawId = os.Getenv("BITBUCKET_WORKSPACE")
+	}
+
+	id, err := deterministic(rawId)
+	if err != nil {
+		return ""
+	}
+
+	return id.String()
+}
+
+func UserId() string {
+	if IsCI() {
+		ci := guessCI()
+		var rawId string
+
+		if ci == "github" {
+			rawId = os.Getenv("GITHUB_ACTOR")
+		} else if ci == "gitlab" {
+			rawId = fmt.Sprintf("%s %s %s",
+				os.Getenv("GITLAB_USER_NAME"),
+				os.Getenv("GITLAB_USER_LOGIN"),
+				os.Getenv("GITLAB_USER_ID"))
+		} else if ci == "bitbucket" {
+			rawId = os.Getenv("BITBUCKET_STEP_TRIGGERER_UUID")
+		} else {
+			var out bytes.Buffer
+			cmd := exec.Command("git", "log", "-1", "--pretty=format:'%ae'")
+			cmd.Stdout = &out
+
+			err := cmd.Run()
+			if err != nil {
+				return ""
+			}
+
+			rawId = out.String()
+		}
+
+		id, err := deterministic(rawId)
+		if err != nil {
+			return ""
+		}
+
+		return id.String()
+	}
+
+	id := uuid.New().String()
+	old := appdirs.UserConfigDir("dvc/user_id", "iterative", "", false)
+	_, errorOld := os.Stat(old)
+
+	new := appdirs.UserConfigDir("iterative/telemetry", "", "", false)
+	_, errorNew := os.Stat(new)
+
+	if os.IsNotExist(errorNew) {
+		if !os.IsNotExist(errorOld) {
+			jsonFile, jsonErr := os.Open(old)
+
+			if jsonErr == nil {
+				byteValue, _ := ioutil.ReadAll(jsonFile)
+				var data map[string]interface{}
+				json.Unmarshal([]byte(byteValue), &data)
+				id = data["user_id"].(string)
+
+				defer jsonFile.Close()
+			}
+		}
+
+		os.MkdirAll(filepath.Dir(new), 0644)
+		ioutil.WriteFile(new, []byte(id), 0644)
+	} else {
+		dat, _ := ioutil.ReadFile(new)
+		id = string(dat[:])
+	}
+
+	if os.IsNotExist(errorOld) {
+		os.MkdirAll(filepath.Dir(old), 0644)
+		data := map[string]interface{}{
+			"user_id": id,
+		}
+		file, _ := json.MarshalIndent(data, "", " ")
+		ioutil.WriteFile(old, file, 0644)
+	}
```

Comment:
ideal pseudo-code for this is meant to be:

```python
if new.exists():
    uid = new.read_text().strip()
else:  # if uid.lower() != "do-not-track":
    if old.exits():
        uid = json.load(old.open())["user_id"]
    else:  # only for non-DVC packages
        # write legacy file in case legacy DVC is installed later
        old.write_text(f'{{"user_id": "{uid}"}}')
    new.write_text(uid)
```

yet instead the current implementation is:

```python
if not new.exists():
    if old.exits():
        uid = json.load(old.open())["user_id"]
    new.write_text(uid)
else:  # new exists - ignore old
    uid = new.read_text().strip()

if not old.exits():
    old.write_text(f'{{"user_id": "{uid}"}}')
```

they're nearly (but not quite) the same.

```suggestion
	if os.IsNotExist(errorNew) {
		if os.IsNotExist(errorOld) {
			os.MkdirAll(filepath.Dir(old), 0644)
			data := map[string]interface{}{
				"user_id": id,
			}
			file, _ := json.MarshalIndent(data, "", " ")
			ioutil.WriteFile(old, file, 0644)
		} else {
			jsonFile, jsonErr := os.Open(old)

			if jsonErr == nil {
				byteValue, _ := ioutil.ReadAll(jsonFile)
				var data map[string]interface{}
				json.Unmarshal([]byte(byteValue), &data)
				id = data["user_id"].(string)

				defer jsonFile.Close()
			}
		}

		os.MkdirAll(filepath.Dir(new), 0644)
		ioutil.WriteFile(new, []byte(id), 0644)
	} else {
		dat, _ := ioutil.ReadFile(new)
		id = string(dat[:])
	}
```

Reply:
@casperdcl your ideal solution is not what we agreed, we were writing the old format in any case with our generated id.

Reply:
@casperdcl

Reply:
hmm ok - updated suggestion. How about now?

Reply:
Include the `do-not-track` exception. 👍

Reply:
@casperdcl that check is not needed. We can not specify the id within the provider. The only way of having `do-not-track` as an id should be reading the new or old credentials generated by other iterative product. If thats the case the file exists and are not going to be written

Reply:
@casperdcl that check is not needed. We can not specify the id within the provider. The only way of having `do-not-track` as an id should be reading the new or old credentials generated by other iterative product. If thats the case the file exists and are not going to be written or the old credentials are going to be written with `do-not-track` which should be fine